File: src/designpatterns/factory/FactoryJava/Burger.java
package designpatterns.factory.FactoryJava;

public abstract class Burger {
    
    abstract void prepare();
    abstract void cook();
    abstract void serve();
}

enum BurgerType {
    CHEESE, DELUXE_CHEESE, VEGAN
}

class CheeseBurger extends Burger {
    
    @Override
    void prepare() {}
    
    @Override
    void cook() {}

    @Override
    void serve() {}
}

class DeluxeCheeseBurger extends Burger {
    
    @Override
    void prepare() {}
    
    @Override
    void cook() {}

    @Override
    void serve() {}
}

class VeganBurger extends Burger {
    
    @Override
    void prepare() {}
    
    @Override
    void cook() {}

    @Override
    void serve() {}
}

File: src/designpatterns/factory/FactoryTS/burger.ts
abstract class Burger {
    abstract prepare(): void;
    abstract cook(): void;
    abstract serve(): void;
}

type BurgerType = 'CHEESE' | 'DELUXE_CHEESE' | 'VEGAN'; // add more Burger type here

class CheeseBurger extends Burger {
    prepare(): void {}

    cook(): void {}

    serve(): void {}
}

class DeluxeCheeseBurger extends Burger {
    prepare(): void {}

    cook(): void {}

    serve(): void {}
}

class VeganBurger extends Burger {
    prepare(): void {}

    cook(): void {}

    serve(): void {}
}

/* and yes... more Burgers = more concrete classes */

export { Burger };
export type { BurgerType };
export { CheeseBurger, DeluxeCheeseBurger, VeganBurger };